crates/aptos-jwk-consensus/src/epoch_manager.rs

```diff
@@ -0,0 +1,225 @@
+            jwk_updated_event_txs: None,
+            jwk_rpc_msg_tx: None,
+            jwk_manager_close_tx: None,
+        }
+    }
+
+    /// On a new RPC request, forward to JWK consensus manager, if it is alive.
+    fn process_rpc_request(
+        &mut self,
+        peer_id: Author,
+        rpc_request: IncomingRpcRequest,
+    ) -> Result<()> {
+        if Some(rpc_request.msg.epoch()) == self.epoch_state.as_ref().map(|s| s.epoch) {
+            if let Some(tx) = &self.jwk_rpc_msg_tx {
+                let _ = tx.push((), (peer_id, rpc_request));
+            }
+        }
+        Ok(())
+    }
+
+    /// On a on-chain JWK updated events, forward to JWK consensus manager if it is alive.
+    fn process_onchain_event(&mut self, notification: EventNotification) -> Result<()> {
+        let EventNotification {
+            subscribed_events, ..
+        } = notification;
+        for event in subscribed_events {
+            if let Ok(jwk_event) = ObservedJWKsUpdated::try_from(&event) {
+                if let Some(tx) = self.jwk_updated_event_txs.as_ref() {
+                    let _ = tx.push((), jwk_event);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn start(mut self, mut network_receivers: NetworkReceivers) {
+        self.await_reconfig_notification().await;
+        loop {
+            let handle_result = tokio::select! {
+                reconfig_notification = self.reconfig_events.select_next_some() => {
+                    self.on_new_epoch(reconfig_notification).await
+                },
+                event = self.jwk_updated_events.select_next_some() => {
+                    self.process_onchain_event(event)
+                },
+                (peer, rpc_request) = network_receivers.rpc_rx.select_next_some() => {
+                    self.process_rpc_request(peer, rpc_request)
+                }
+            };
+
+            if let Err(e) = handle_result {
+                error!("{}", e);
+            }
+        }
+    }
+
+    async fn await_reconfig_notification(&mut self) {
+        let reconfig_notification = self
+            .reconfig_events
+            .next()
+            .await
+            .expect("Reconfig sender dropped, unable to start new epoch");
+        self.start_new_epoch(reconfig_notification.on_chain_configs)
+            .await;
+    }
+
+    async fn start_new_epoch(&mut self, payload: OnChainConfigPayload<P>) {
+        let validator_set: ValidatorSet = payload
+            .get()
+            .expect("failed to get ValidatorSet from payload");
+
+        let epoch_state = Arc::new(EpochState {
+            epoch: payload.epoch(),
+            verifier: (&validator_set).into(),
+        });
+        self.epoch_state = Some(epoch_state.clone());
+        info!("[JWK] start_new_epoch: new_epoch={}", epoch_state.epoch);
+
+        let features = payload.get::<Features>().unwrap_or_default();
+
+        if features.is_enabled(FeatureFlag::JWK_CONSENSUS) {
```

Comment:
Also need to check vtxn is enabled?

Reply:
i still think we don't have to.
(Even for DKG i want to change back if we can).
1-to-1 mapping between flags and components would be ideal.

Reply:
what happens if vtxn is not enabled? just consensus wont' pull vtxn?

Reply:
Then how do we ensure correctness? @zekun000 for discussion as well

Reply:
yes, consensus vtxn config controlls if consensus pull vtxns.
jwk feature controls if jwk vtxns are produced. (we have TxnGuard to ensure txns are dropped at the end of epoch)

Reply:
nothing will break?

Reply:
let's say nothing should break :)

Reply:
Can we add a smoke test for this? :)

Reply:
good idea.